File: Codility/7_stacks_and_queues/stone_wall.py
# https://codility.com/programmers/task/stone_wall/

def solution(H):
    count = 0
    stack = []
    for h in H:
        while stack and stack[-1] > h:
            stack.pop()
            
        if not stack or stack[-1] < h:
            count += 1
            stack.append(h)
            
    return count

File: Codility/7_stacks_and_queues/stone_wall.cpp
// https://codility.com/programmers/task/stone_wall/

int solution(vector<int> &H) {
    vector<int> stack;
    int count = 0;
    for (int h : H)
    {
        while (!stack.empty() && h < stack.back())
            stack.pop_back();

        if (stack.empty() || h > stack.back())
        {
            stack.push_back(h);
            ++count;
        }
    }
    
    return count;
}